src/google-classroom-extractor/google_classroom_extractor/api/courses.py

```diff
@@ -119,18 +121,180 @@ def request_all_courses_as_df(
     """
 
     courses_df = request_latest_courses_as_df(resource)
+    _sync_courses(courses_df, sync_db)
+    return courses_df
+
+
+def _sync_courses(courses_df: DataFrame, sync_db: sqlalchemy.engine.base.Engine):
+    """
+    Take fetched API data and sync with database.
+
+    Parameters
+    ----------
+    courses_df: DataFrame
+        a Courses API DataFrame with the current fetched data which
+        will be mutated, adding Hash and CreateDate/LastModifiedDate
+    sync_db: sqlalchemy.engine.base.Engine
+        an Engine instance for creating database connections
+    """
+    courses_df.set_index('id', inplace=True)
+
+    # compute hash from API call data
```

Comment:
create a hash for each row

Reply:
You might be interested in in an alternative built right into Pandas... see [this notebook](https://github.com/Ed-Fi-Exchange-OSS/Ed-Fi-X-Fizz/blob/09c668c3435512688925638d10a45da5bd964b76/experimental/python/hashing.ipynb) I just uploaded.

Reply:
Ok, sold thanks to the speed.  As long as it's stable across application launches.  The Python hash() function is not, and hash_pandas_object() is poorly documented. I'd seen hash_pandas_object() but didn't know if it used hash() under the hood, and had seen that it has trouble with some data types, probably not relevant though.

Reply:
It turned out that hash_pandas_object() is not always stable after all.  However, in playing around with pandas row serialization I found that to_json() is way faster than to_csv().  Switching to that plus switching to a popular non-crypto hash library - xxhash - gives much improved performance.  Added perf test to notebook.